packages/core/src/services/chains/run.ts

```diff
@@ -103,175 +114,63 @@ async function iterate({
   configOverrides?: ConfigOverrides
 }) {
   const prevText = previousResponse?.text
-  const stepValidator = new ChainValidator({ prevText, chain, providersMap })
-  const step = await stepValidator.call().then((r) => r.unwrap())
+  const chainValidator = new ChainValidator({
+    workspace,
+    prevText,
+    chain,
+    providersMap,
+    configOverrides,
+  })
+  const streamConsumer = new ChainStreamConsumer({
+    controller,
+    previousCount,
+    documentLogUuid,
+  })
 
   try {
-    const newMessagesInStep = step.conversation.messages.slice(previousCount)
-    const sentCount = previousCount + newMessagesInStep.length
-    publishStepStartEvent(controller, step, newMessagesInStep, documentLogUuid)
-
+    const step = await chainValidator.call().then((r) => r.unwrap())
     const providerProcessor = new ProviderProcessor({
       source,
       documentLogUuid,
       config: step.config,
       apiProvider: step.provider,
       messages: step.conversation.messages,
+      saveSyncProviderLogs: step.chainCompleted,
     })
+    const { messageCount, stepStartTime } = streamConsumer.start({ step })
```

Comment:
object argument with a field looks confusing to me i would expect more than one field

Reply:
I don't understand this comment

Reply:
`streamConsumer.start(step)`

Reply:
Ahh! this discussion xD

Reply:
Changed